### ISSUE TITLE
[Feature Request] tf.resume_gradient

### ISSUE THREAD

**Issue by ybsave** [tensorflow/tensorflow#14162]
tf.stop_gradient provides a convenient way for various uses. However, sometimes, it may be useful to resume the gradient passing after some conditions are met. 

For example, in Faster-RCNN algorithm, the RPN layers' gradients are stopped because they may cause unstable results in early rounds. However, this indicates that these layers' weights would be kept the initial random values forever. If we could resume the gradients back propagation after several steps, the results may be better. Therefore, I believe that the resume_gradient function would be useful. Thank you.

**Comment by ppwwyyxx** [comment]
`tf.where(condition, x, tf.stop_gradient(x)`

**Comment by jart** [comment]
@ybsave Does @ppwwyyxx's suggestion satisfy your needs? (Thank you by the way @ppwwyyxx)

**Comment by ybsave** [comment]
@ppwwyyxx Thank you. I think this is OK for my use.

**Comment by ybsave** [comment]
Hello, I just made some simple tests and find that the tf.where may not satisfy my need. I want the gradient of a loss to stop back propagation in early rounds, but continue the BP in later rounds. 

As loss is usually defined before a session is really created, if I use codes like
    loss = tf.where(math_ops.greater_equal(global_step, STEP_THRESHOLD), loss, tf.stop_gradient(loss))

I found that the gradient still passes when global_step > STEP_THRESHOLD. This does not meet my requirement. 

Is there anyway to do this (using global_step to control "stop_gradient")? Or a new resume_gradient feature is still needed? Thank you.

**Comment by ppwwyyxx** [comment]
> I found that the gradient still passes when global_step > STEP_THRESHOLD. This does not meet my requirement.

This is exactly what you wrote. Gradient doesn't pass only when step <= STEP_THRESHOLD.

**Comment by ybsave** [comment]
@ppwwyyxx I am sorry for my typo. I have corrected it. The gradient still pass when global_step < STEP_THRESHOLD.

**Comment by ppwwyyxx** [comment]
You'll need to give a small repro then, for bugs like this. At least it works in my test:
```python

**Comment by ybsave** [comment]
@ppwwyyxx Thank you so much for your quick response. To clarify my question, I am using slim.learning.train to perform the training, where the global_step increment was performed inside the function.

However, when I define my loss, it is defined before any session is created. Thus, when I use the global_step variable, I cannot get the correct value as tf.train.global_step requires a session as input. If I just use tf.where(tf.greater(global_step, 2), loss, tf.stop_gradient(loss)) instead of your codes like tf.where(tf.greater(gs_inc, 2), loss, tf.stop_gradient(loss)), I cannot get the correct global_step. Therefore, I wonder whether there is any method to perform the tf.where for defining the loss without using operators like "gs_inc = tf.assign_add(gs, 1)"?

I am currently testing it in a big program. I will try to write a small program to show the problem soon. I appreciate your kind help.

**Comment by ppwwyyxx** [comment]
Though I don't think I fully understand what's your problem, but you should not use `tf.train.global_step` for this. `tf.train.get_or_create_global_step` will give you the global step __tensor__ (instead of the value) and that's what you need. But still I'm not sure whether it works with high-level wrappers such as slim which takes control of the increment.

**Comment by ybsave** [comment]
Hello, I made a simple test as below

	  TRAIN_STEP_NUM = 5
	  x_train = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
	  y_train = [0, -1, -2, -3, -4, -5, -6, -7, -8, -9]

	  dataset = tf.data.Dataset.from_tensor_slices(
		  (tf.cast(x_train, tf.float32), tf.cast(y_train, tf.float32)))

	  with tf.Graph().as_default():
		iterator = dataset.batch(1).make_one_shot_iterator()
		x, y = iterator.get_next()
		
		W = tf.Variable([.3], dtype=tf.float32)
		b = tf.Variable([-.3], dtype=tf.float32)
		linear_model = W*x_train + b
		loss = tf.reduce_sum(tf.square(linear_model - y))

		loss = tf.where(
		  tf.greater(tf.train.get_or_create_global_step(), TRAIN_STEP_NUM),
					 loss, tf.stop_gradient(loss))
		slim.losses.add_loss(loss)
	  
		optimizer = tf.train.GradientDescentOptimizer(0.01)
		train_op = slim.learning.create_train_op(loss, optimizer)

		slim.learning.train(train_op, './tmp',
							log_every_n_steps=1,
							number_of_steps=10)

The outputs are:

	INFO:tensorflow:Starting Session.
	INFO:tensorflow:global_step/sec: inf
	INFO:tensorflow:Saving checkpoint to path ./tmp\model.ckpt
	INFO:tensorflow:Starting Queues.
	INFO:tensorflow:global step 1: loss = 25.6500 (0.016 sec/step)
	INFO:tensorflow:global step 2: loss = 62.6500 (0.000 sec/step)
	INFO:tensorflow:global step 3: loss = 119.6500 (0.000 sec/step)
	INFO:tensorflow:global step 4: loss = 196.6500 (0.000 sec/step)
	INFO:tensorflow:global step 5: loss = 293.6500 (0.000 sec/step)
	INFO:tensorflow:global step 6: loss = 410.6500 (0.000 sec/step)
	INFO:tensorflow:global step 7: loss = 547.6500 (0.000 sec/step)
	INFO:tensorflow:global step 8: loss = 21905.0312 (0.000 sec/step)
	INFO:tensorflow:global step 9: loss = 1032078.1250 (0.000 sec/step)
	INFO:tensorflow:global step 10: loss = 48496848.0000 (0.000 sec/step)
	INFO:tensorflow:Stopping Training.
	INFO:tensorflow:Finished training! Saving model to disk.

There must be wrong in my codes, but I do not know why. 1. The loss is not reducing; 2) the stop_gradient did not freeze the updating of parameters, leading to loss change. 

Would you please point out my error? Thank you.

**Comment by ppwwyyxx** [comment]
The loss is changing because you're sending different data in each iteration. You snippet doesn't show any bugs.

**Comment by ybsave** [comment]
@ppwwyyxx  Thank you for your kind help. After using the same data, the program now performs as expected. Really appreciate!
